src/Data/Nat/Primality/Factorisation.agda

```diff
@@ -0,0 +1,218 @@
+  pattern 2≤2+n = s≤s (s≤s z≤n)
+  pattern 1<2+n = 2≤2+n
+
+factorise : ∀ n → .{{NonZero n}} → PrimeFactorisation n
+factorise 1 = record
+  { factors = []
+  ; isFactorisation = refl
+  ; factorsPrime = []
+  }
+factorise (2+ n) = <-rec P factoriseRec (2 + n) {2} 2≤2+n (≤⇒≤‴ 2≤2+n) (2-rough-n (2 + n))
+  where
+
+  P : ℕ → Set
+  P n′ = ∀ {k} → 2 ≤ n′ → k ≤‴ n′ → k Rough n′ → PrimeFactorisation n′
+
+  factoriseRec : ∀ n → <-Rec P n → P n
+  factoriseRec (2+ n) rec (s≤s (s≤s n≤z)) ≤‴-refl k-rough-n = record
+    { factors = 2 + n ∷ []
+    ; isFactorisation = *-identityʳ (2 + n)
+    ; factorsPrime = k-rough-n ∷ []
+    }
+  factoriseRec (2+ n) rec {0} 2≤2+n (≤‴-step (≤‴-step k<n)) k-rough-n =
+    factoriseRec (2+ n) rec 2≤2+n k<n (2-rough-n (2+ n))
+  factoriseRec (2+ n) rec {1} 2≤2+n (≤‴-step k<n) k-rough-n =
+    factoriseRec (2+ n) rec 2≤2+n k<n (2-rough-n (2+ n))
+  factoriseRec (2+ n) rec {suc (suc k)} 2≤2+n (≤‴-step k<n) k-rough-n with 2 + k ∣? 2+ n
+  ... | no  k∤n = factoriseRec (2+ n) rec {3 + k} 2≤2+n k<n (extend-∤ k-rough-n k∤n)
+  ... | yes k∣n = record
+    { factors = 2 + k ∷ factors res
+    ; isFactorisation = prop
+    ; factorsPrime = roughn∧∣n⇒prime k-rough-n 2≤2+n k∣n ∷ factorsPrime res
+    }
+    where
+      open ≤-Reasoning
+
+      q : ℕ
+      q = quotient k∣n
+
+      -- we know that k < n, so if q is 0 or 1 then q * k < n
+      2≤q : 2 ≤ q
+      2≤q = ≮⇒≥ (λ q<2 → contradiction (_∣_.equality k∣n) (>⇒≢ (prf (≤-pred q<2)))) where
+
+        prf : q ≤ 1 → q * 2+ k < 2 + n
+        prf q≤1 = begin-strict
+          q * 2+ k  ≤⟨ *-monoˡ-≤ (2 + k) q≤1 ⟩
+          2 + k + 0 ≡⟨ +-identityʳ (2 + k) ⟩
+          2 + k     <⟨ ≤‴⇒≤ k<n ⟩
+          2 + n     ∎
+
+      0<q : 0 < q
+      0<q = begin-strict
+        0 <⟨ s≤s z≤n ⟩
+        2 ≤⟨ 2≤q ⟩
+        q ∎
+
+      q<n : q < 2 + n
+      q<n = begin-strict
+        q           <⟨ m<m*n q (2 + k) ⦃ >-nonZero 0<q ⦄ 2≤2+n ⟩
+        q * (2 + k) ≡˘⟨ _∣_.equality k∣n ⟩
+        2 + n       ∎
+
+      q≮2+k : q ≮ 2 + k
+      q≮2+k q<k = k-rough-n 2≤q q<k (quotient∣n k∣n)
+
+      res : PrimeFactorisation q
+      res = rec q q<n {2 + k} 2≤q (≤⇒≤‴ (≮⇒≥ q≮2+k))
+          $ λ {d} d<k d-prime → k-rough-n d<k d-prime ∘ flip ∣-trans (quotient∣n k∣n)
+
+      prop : (2 + k) * product (factors res) ≡ 2 + n
+      prop = begin-equality
+        (2 + k) * product (factors res)
+          ≡⟨ cong ((2 + k) *_) (isFactorisation res) ⟩
+        (2 + k) * q
+          ≡⟨ *-comm (2 + k) q ⟩
+        q * (2 + k)
+          ≡˘⟨ _∣_.equality k∣n ⟩
+        2 + n ∎
+
+------------------------------------------------------------------------
+-- Properties of a factorisation
+
+factorisation≥1 : ∀ {as} → All Prime as → product as ≥ 1
+factorisation≥1 {[]} [] = s≤s z≤n
+factorisation≥1 {suc a ∷ as} (pa ∷ asPrime) = *-mono-≤ {1} {1 + a} (s≤s z≤n) (factorisation≥1 asPrime)
+
+factorisationPullToFront : ∀ {as} {p} → Prime p → p ∣ product as → All Prime as → ∃[ as′ ] as ↭ (p ∷ as′)
```

Comment:
Wouldn't it be 'simpler' if this took a `PrimeFactorisation` as input instead of the pieces it contains?

Also it feels like this could be proven more simply:
1. prove p is a member of the list as
2. then there's a permutation that puts it at the front.

This would be straightforward if you already knew it was a PrimeFactorisation.

Reply:
I've simplified it a bit. I don't think I do want it to take a `PrimeFactorisation`, because it recurses on the list, which would be a little more awkward. But maybe I'm wrong here, I'll think about it more